File: BTFromInorderAndPostorder.py
# Time complexity: O(n) where n is the number of elements in inorder and postorder array
# Space complexity: O(n)

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder, postorder):
        if len(inorder) == 0 or len(postorder) == 0:
            return
        nodeVal = postorder.pop()
        root = TreeNode(nodeVal)
        inorderIndex = inorder.index(nodeVal)
        root.right = self.buildTree(inorder[inorderIndex+1:], postorder)
        root.left = self.buildTree(inorder[:inorderIndex], postorder)
        return root

File: BTfromInOrderAndPostOrder.cpp
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> mp; 
        for (int i = 0; i < inorder.size(); ++i) mp[inorder[i]] = i; 
        
        stack<TreeNode*> stk; 
        TreeNode *root = nullptr, *node = nullptr; 
        for (int i = postorder.size()-1; i >= 0; --i) {
            int x = postorder[i]; 
            if (!root) root = node = new TreeNode(x); 
            else if (mp[node->val] < mp[x]) {
                stk.push(node); 
                node = node->right = new TreeNode(x); 
            } else {
                while (stk.size() && mp[x] < mp[stk.top()->val]) node = stk.top(), stk.pop(); 
                node = node->left = new TreeNode(x); 
            }
        }
        return root;
    }
};
